### PR TITLE
35% faster packing + rename `bfd-requeue` to `bfd_split`

### PR DESCRIPTION
**Pull request by mariosasko**
# What does this PR do?

This PR improves the packing logic to make it faster, less error-prone, and easier to read.

The main changes are:

* Replacing the AI-generated BFD splitting (a.k.a. "requeuing") logic from https://github.com/huggingface/trl/pull/4632
 with a vectorized implementation that is significantly shorter and 30% faster.

* Updating `pack_examples` to restore the input dataset’s format and perform proper input validation.

* Applying a minor optimization to the `wrapped` packing implementation by reusing the `offsets` across all packed columns.

* Aligning the naming with the literature (e.g., `requeue` → `split`) while preserving backward compatibility.

P.S. The recent [Qwen3-Coder-Next technical report](https://github.com/QwenLM/Qwen3-Coder/blob/main/qwen3_coder_next_tech_report.pdf) includes a nice comparison of packing techniques, which would be a nice addition to the docs. ~~However, the report is not yet available on arXiv, so it cannot be cited as an HF paper.~~ Done 😄 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Touches core SFT data preprocessing (packing/truncation) and changes strategy semantics/validation, which can alter token ordering/retention and training results if misconfigured; backward compatibility is handled via a deprecation warning.
> 
> **Overview**
> **Improves SFT dataset packing and clarifies packing strategies.** The BFD packer is rewritten with a more vectorized PyArrow implementation, adds shared column validation, filters empty sequences, and introduces an explicit overflow behavior switch (*truncate* vs *split*) used by the new `"bfd_split"` strategy.
> 
> **API/UX updates around packing.** `pack_dataset` now validates strategies, preserves/restores the original dataset format (including IterableDatasets) and ensures `seq_lengths` is included in formatted columns for BFD-based strategies; `truncate_dataset` is unified to the same Arrow-based map + format restore path.
> 
> **Naming + docs/tests alignment.** `SFTConfig` deprecates `bfd-requeue` by warning and mapping it to `bfd_split`, `SFTTrainer` treats `bfd_split` like `bfd` for padding-free/FlashAttention checks, tests are updated to assert format preservation and `bfd_split` behavior (including empty sequences), and docs add/expand guidance and citations for `bfd_split` vs `wrapped`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b265bbf4c95e38f1da81475847ac83b420f3fd27. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: trl/data_utils.py [hunk 1]
```diff
@@ -612,6 +612,29 @@ def maybe_extract_prompt(example: dict[str, list]) -> dict[str, list]:
...
+
+def _check_if_columns_can_be_packed(columns: list[pa.Array]):
+    first_column_offsets = None
+    for idx, column in enumerate(columns):
+        if not (pyarrow.types.is_list(column.type) or pyarrow.types.is_large_list(column.type)):
+            raise TypeError("Packing requires all columns to be lists of lists.")
+
+        if idx == 0:
+            first_column_offsets = column.offsets
+        elif first_column_offsets != column.offsets:
```
[the remark this answers is not in the document]
Comment (mariosasko) [reply trl/data_utils.py]: PyArrow Array's `__eq__` [calls](https://github.com/apache/arrow/blob/dbbf7cf8136da31b77f78f00a51d62992802c2c2/python/pyarrow/array.pxi#L1473) `.equals()`, so the result is a scalar `bool`, not a boolean array like in NumPy/PyTorch.

But I still switched to `.equals()` to make the bot happy :)

File: trl/data_utils.py [hunk 2]
```diff
@@ -855,27 +839,27 @@ def pack_dataset(
...
+            fn_kwargs={"seq_length": seq_length, "on_seq_length_overflow": "split"},
             **map_kwargs,
         )
     elif strategy == "wrapped":
         dataset = dataset.map(_pack_wrapped, batched=True, fn_kwargs={"seq_length": seq_length}, **map_kwargs)
     else:
-        raise ValueError(f"Invalid packing strategy: {strategy}. Use 'bfd', 'bfd-requeue', or 'wrapped'.")
-    dataset = dataset.with_format(None)
+        raise ValueError(f"Invalid packing strategy: {strategy}. Use 'bfd', 'bfd-split', or 'wrapped'.")
+    dataset = dataset.with_format(**format)
```
[the remark this answers is not in the document]
Comment (mariosasko) [reply trl/data_utils.py:876]: This is a really nice catch! The problematic pattern in question comes from `datasets` (e.g. [here](https://github.com/huggingface/datasets/blob/6f9d24f160c7fc92836192d9092556a252477e10/src/datasets/arrow_dataset.py#L2414-L2429)), so I made the necessary changes (+ a test) to address it.

File: docs/source/data_utils.md [hunk 3]
```diff
@@ -15,3 +15,15 @@
...
+
+[[autodoc]] truncate_dataset
+
+## pack_dataset
+
+[[autodoc]] pack_dataset
+
+## PackingStrategy
+
+[[autodoc]] PackingStrategy
```
Comment (qgallouedec) [inline docs/source/data_utils.md]:
```suggestion
```


see https://github.com/huggingface/trl/issues/5090

File: trl/data_utils.py [hunk 4]
```diff
@@ -15,20 +15,48 @@
...
+                "bfd_requeue": "bfd_split",
+            }
+            if normalized in aliases:
+                return cls(aliases[normalized])
+
+        # Copied from https://github.com/huggingface/transformers/blob/1a50a3b13b6d17c2637fe19e94a8c459bd4208a5/src/transformers/utils/generic.py#L485-L487
+        raise ValueError(
+            f"{value} is not a valid {cls.__name__}, please select one of {list(cls._value2member_map_.keys())}"
+        )
+
```
Comment (qgallouedec) [inline trl/data_utils.py]: ideally we want to keep things simple, ie not use enum when possible

File: trl/trainer/sft_config.py [hunk 5]
```diff
@@ -190,13 +191,13 @@ class SFTConfig(_BaseConfig):
             "and reduce padding. Uses `max_length` to define sequence length."
         },
     )
-    packing_strategy: str = field(
+    packing_strategy: str | PackingStrategy = field(
         default="bfd",
         metadata={
             "help": "Strategy for packing sequences. Can be `'bfd'` (best-fit decreasing, truncates overflow), "
-            "`'bfd-requeue'` (best-fit decreasing, re-queues overflow tokens), or `'wrapped'` (aggressive, cuts "
+            "`'bfd_split'` (best-fit decreasing, splits overflow sequences), or `'wrapped'` (aggressive, cuts "
```
Comment (qgallouedec) [inline trl/trainer/sft_config.py:196]: > Aligning the naming with the literature (e.g., requeue → split) while preserving backward compatibility.

nice! I quickly checked the Qwen3-Coder-Tech report, what that call "concat-then-split" is closer to the "wrapped" strategy no?

File: trl/data_utils.py [hunk 6]
```diff
@@ -657,75 +709,37 @@ def search(self, val):
...
+    if on_seq_length_overflow == "truncate":
+        columns = [pc.list_slice(column, 0, seq_length) for column in columns]
+    elif on_seq_length_overflow == "split":
+        lengths = pc.list_value_length(columns[0]).to_numpy()
+        # Split the sequences longer than `seq_length` into chunks (of length `seq_length` or less) while respecting sequence boundaries
+        num_fragments = np.ceil(lengths / seq_length).astype(int)
+        offsets = np.arange(np.sum(num_fragments) + 1, dtype=columns[0].offsets.type.to_pandas_dtype()) * seq_length
+        # "Left-shift" the offsets to account for the last fragment of each original sequence possibly being shorter than `seq_length`
+        diff = np.zeros_like(offsets)
+        diff[np.cumsum(num_fragments)] = -lengths % seq_length
```
[the remark this answers is not in the document]
Comment (mariosasko) [reply trl/data_utils.py:708]: Also decided to address this one, even though empty examples are not expected.

File: trl/data_utils.py (hunk 4 above)
Comment (mariosasko) [reply trl/data_utils.py]: Some extra code to make aliasing for backward compatibility more elegant/less error-prone is worth it IMO.

PS: The aliasing logic also replaces `-` with `_` to align with the config parameter naming used in Transformers/TRL.

File: trl/trainer/sft_config.py (hunk 5 above)
Comment (mariosasko) [reply trl/trainer/sft_config.py:196]: Yes. I think [Fewer Truncations Improve Language Modeling](https://arxiv.org/abs/2404.10830) was the first one to use this terminology.

Do you want me to alias "wrapped" to "concat-then-split" then?

File: trl/data_utils.py [hunk 7]
```diff
@@ -855,27 +874,29 @@ def pack_dataset(
...
             **map_kwargs,
         )
-    elif strategy == "wrapped":
+    else:  # PackingStrategy.WRAPPED
         dataset = dataset.map(_pack_wrapped, batched=True, fn_kwargs={"seq_length": seq_length}, **map_kwargs)
-    else:
-        raise ValueError(f"Invalid packing strategy: {strategy}. Use 'bfd', 'bfd-requeue', or 'wrapped'.")
-    dataset = dataset.with_format(None)
+    if strategy in {PackingStrategy.BFD, PackingStrategy.BFD_SPLIT} and "columns" in format:
+        format["columns"].append("seq_lengths")
```
[the remark this answers is not in the document]
Comment (mariosasko) [reply trl/data_utils.py]: The `"columns"` key cannot be `None` (see [here](https://github.com/huggingface/datasets/blob/36b79a66fdad4195a78d281214ed2d13c625fbbf/src/datasets/arrow_dataset.py#L2805))

**Comment by qgallouedec** [comment]
Thanks @mariosasko reviewing it today

File: trl/data_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mariosasko) [reply trl/data_utils.py]: Good call! `datasets` should return a copy to guard against in-place modifications, but that's not the case.

File: trl/data_utils.py (hunk 4 above)
Comment (qgallouedec) [reply trl/data_utils.py]: > PS: The aliasing logic also replaces - with _ to align with the config parameter naming used in Transformers/TRL.

Sounds good!

> Some extra code to make aliasing for backward compatibility more elegant/less error-prone is worth it IMO

fair, I've addressed this directly in the post-init of the config, that's usually how we do in the repo

File: trl/trainer/sft_config.py (hunk 5 above)
Comment (qgallouedec) [reply trl/trainer/sft_config.py:196]: for now I think it's ok to keep "wrapped". A clear documentation is enough IMO. Plus we don't want to advertise to much to packing strategy aha

**Comment by qgallouedec** [comment]
```python
import gc
import time

import numpy as np
from datasets import Dataset

from trl import pack_dataset


def generate_dataset(num_examples, mean_length=256, std_length=128, seed=42):
    """Generate a synthetic dataset with variable-length sequences."""
    rng = np.random.default_rng(seed)
    lengths = rng.normal(mean_length, std_length, num_examples).astype(int).clip(min=1)
    input_ids = [rng.integers(0, 32000, size=length).tolist() for length in lengths]
    attention_mask = [[1] * length for length in lengths]
    return Dataset.from_dict({"input_ids": input_ids, "attention_mask": attention_mask})


def benchmark(num_examples, seq_length, strategy, min_runs=5, min_time=2.0, **kwargs):
    dataset = generate_dataset(num_examples, **kwargs)
    # Warmup (3 runs to stabilize caches)
    for _ in range(3):
        pack_dataset(dataset, seq_length=seq_length, strategy=strategy)

    # Run until we have enough samples and enough wall time
    times = []
    total_elapsed = 0.0
    while len(times) < min_runs or total_elapsed < min_time:
        gc.collect()
        gc.disable()
        start = time.perf_counter()
        packed = pack_dataset(dataset, seq_length=seq_length, strategy=strategy)
        elapsed = time.perf_counter() - start
        gc.enable()
        times.append(elapsed)
        total_elapsed += elapsed

    # Use median instead of mean to reduce sensitivity to outliers
    median_time = np.median(times)
    iqr = np.percentile(times, 75) - np.percentile(times, 25)
    total_tokens = sum(len(ids) for ids in dataset["input_ids"])
    packed_examples = len(packed)

    return {
        "num_examples": num_examples,
        "seq_length": seq_length,
        "strategy": strategy,
        "total_tokens": total_tokens,
        "packed_examples": packed_examples,
        "median_time": median_time,
        "iqr": iqr,
        "num_runs": len(times),
        "throughput_examples_per_sec": num_examples / median_time,
        "throughput_tokens_per_sec": total_tokens / median_time,
    }


if __name__ == "__main__":
    configs = [
        # Vary dataset size
        {"num_examples": 1_000, "seq_length": 512, "strategy": "bfd"},
        {"num_examples": 10_000, "seq_length": 512, "strategy": "bfd"},
        {"num_examples": 100_000, "seq_length": 512, "strategy": "bfd"},
        # Vary seq_length
        {"num_examples": 10_000, "seq_length": 256, "strategy": "bfd"},
        {"num_examples": 10_000, "seq_length": 1024, "strategy": "bfd"},
        {"num_examples": 10_000, "seq_length": 4096, "strategy": "bfd"},
        # Compare strategies
        {"num_examples": 10_000, "seq_length": 512, "strategy": "bfd_split"},
        {"num_examples": 10_000, "seq_length": 512, "strategy": "wrapped"},
    ]

    header = f"{'strategy':<12} {'examples':>10} {'seq_len':>8} {'packed':>8} {'runs':>5} {'median (s)':>14} {'tok/s':>14}"
    print(header)
    print("-" * len(header))

    for cfg in configs:
        result = benchmark(**cfg)
        print(
            f"{result['strategy']:<12} "
            f"{result['num_examples']:>10,} "
            f"{result['seq_length']:>8} "
            f"{result['packed_examples']:>8,} "
            f"{result['num_runs']:>5} "
            f"{result['median_time']:>8.3f} ± {result['iqr']:.3f} "
            f"{result['throughput_tokens_per_sec']:>14,.0f}"
        )
```


benchmark_results

Nice!!

TBH, I didn't look deeply into the new logic. My understanding is pretty basic, I get the general idea of how it works, I don't see any obvious mistakes, the tests are passing, and the agents are happy. So I'm going to approve and merge this. Albert, feel free to take a look too later; you have more experience with datasets than I do.

File: trl/data_utils.py [hunk 8]
```diff
@@ -855,27 +845,35 @@ def pack_dataset(
     """
     if map_kwargs is None:
         map_kwargs = {}
-    # Fast packing with pyarrow
+
+    valid_strategies = ("bfd", "bfd_split", "wrapped")
+    if strategy not in valid_strategies:
+        raise ValueError(f"Invalid packing strategy '{strategy}', must be one of {valid_strategies}.")
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/data_utils.py:851]: `pack_dataset` ain't part of the public api
